### DISCUSSION TITLE
How to initialValues with typescript, SSR

### DISCUSSION THREAD

**Discussion by leedoopal** [pmndrs/jotai#1500] [Q&A] [answered]
Hi! I'm using jotai with next.js and typescript.

It is a form that uses useHydrateAtoms to put a value in the atom.
I'm looking for advice on what to do when specifying types.

```typescript
// pages/index.tsx

const Page = ({ initialStates }) => {
  useHydrateAtoms([
    [userAtom, initialStates.user],
  ] as const);
}
```

```typescript
// store
export const userAtom = atom<UserModel>({
  name: '',
  info: {
    grade: '',
    isAdult: false
  }
})
```
Wouldn't it be better to put a value deep into the object like this?

Or is it better to set it to null?
```typescript
export const userAtom = atom<UserModel | null>(null);
```

Does using hydrateAtom change all the values in initialValues ?
Or is it a way to update only the values that change?

I'm not sure how to specify the type and how to input initialValues.

**Comment by Thisen** [comment, marked as answer]
`useHydrateAtoms` rewrites the whole value for the given atom. It's meant to be used when the value you want to store, comes from the server (like server side props in Next.js) and you want to supply that to the atom.

If you want to do something partially, you'd need to drill a prop down through the component tree with the value or you could derive the atom value from another value like this:
```ts
const hydratedUserAtom = atom({ ... });
const userAtom = atom((get) => {
  const hydratedUser = get(hydratedUserAtom)
  // Do some transformation
}
```

**Comment by leedoopal** [comment]
@Thisen 
Thanks for letting me know.

An additional question is how best to use the typescript.
Wouldn't it be better to specify an object to the depths? 
```typescript
// store
export const userAtom = atom<UserModel>({
  name: '',
  info: {
    grade: '',
    isAdult: false
  }
})
```

Or is it better to leave it as Null?
```typescript
export const userAtom = atom<UserModel | null>(null);
```

**Reply by Thisen** [reply]
I would say it depends on your requirements. Can your application be used unauthenticated? If so, `null` sounds good.

**Reply by leedoopal** [reply]
Oh thanks! it was helpful
